hibernate-core/src/main/java/org/hibernate/query/results/DomainResultCreationStateImpl.java

```diff
@@ -326,28 +333,47 @@ public DynamicFetchBuilderLegacy resolve(String ownerTableAlias, String fetchedP
 
 		final Consumer<Fetchable> fetchableConsumer = fetchable -> {
 			final String fetchableName = fetchable.getFetchableName();
-			final NavigablePath fetchPath = fetchParent.resolveNavigablePath( fetchable );
-			final NavigablePath relativePath = relativePathStack.isEmpty()
-					? new NavigablePath( fetchableName )
-					: relativePathStack.getCurrent().append( fetchableName );
+			Map.Entry<String, NavigablePath> currentEntry;
+			if ( relativePathStack.isEmpty() ) {
+				currentEntry = new AbstractMap.SimpleEntry<>(
+						getRelativePath( "", fetchable, fetchableContainer ),
+						new NavigablePath( fetchableName )
+				);
+			}
+			else {
+				final Map.Entry<String, NavigablePath> oldEntry = relativePathStack.getCurrent();
+				final String key = oldEntry.getKey();
+				currentEntry = new AbstractMap.SimpleEntry<>(
+						getRelativePath( !key.isEmpty() ? key : "", fetchable, fetchableContainer ),
+						oldEntry.getValue().append( fetchableName )
+				);
+			}
 			// todo (6.0): figure out if we can somehow create the navigable paths in a better way
+			final String fullPath = currentEntry.getKey();
+			FetchBuilder explicitFetchBuilder = fetchBuilderResolverStack
+					.getCurrent()
+					.apply( fullPath );
 			if ( fetchable instanceof Association && fetchable.getMappedFetchOptions().getTiming() == FetchTiming.DELAYED ) {
 				final Association association = (Association) fetchable;
 				final ForeignKeyDescriptor foreignKeyDescriptor = association.getForeignKeyDescriptor();
-				if ( foreignKeyDescriptor.getPartMappingType() instanceof EmbeddableMappingType ) {
-					relativePathStack.push( relativePath.append( ( (EmbeddableMappingType) foreignKeyDescriptor.getPartMappingType() ).getPartName() ) );
-				}
-				else {
-					relativePathStack.push( relativePath.append( foreignKeyDescriptor.getPartName() ) );
+
+				final String partName = attributeName( foreignKeyDescriptor.getSide( association.getSideNature().inverse() )
+						.getModelPart());
+
+				if ( explicitFetchBuilder == null && partName != null ) {
+					currentEntry = new AbstractMap.SimpleEntry<>(
+							currentEntry.getKey() + "." + partName,
+							currentEntry.getValue().append( partName )
+					);
+					explicitFetchBuilder = fetchBuilderResolverStack
+							.getCurrent()
+							.apply( currentEntry.getKey() );
```

Comment:
Not sure if we actually have a test where any of this returns a fetch builder, but I believe that whatever fetch builder we get, it is not the one that we should actually use for this fetchable, as that would represent the FK. Either way, there is something missing here:

```suggestion
					explicitFetchBuilder = fetchBuilderResolverStack
							.getCurrent()
							.apply( currentEntry.getKey() );
					if ( explicitFetchBuilder == null ) {
						fetchBuilderLegacy = legacyFetchResolver.resolve(
							fromClauseAccess.findTableGroup( fetchParent.getNavigablePath() )
									.getPrimaryTableReference()
									.getIdentificationVariable(),
							fetchableName + "." + partName
						);
					}
```

But maybe it's best if we try to write a test case that hits this so we can see what happens. A `@ColumnResult(name = "captain.id"` or something should show the issue in case we actually fetch data.